server/src/main/java/org/elasticsearch/index/store/FsDirectoryFactory.java

```diff
@@ -109,12 +116,28 @@ public static boolean isHybridFs(Directory directory) {
         return unwrap instanceof HybridDirectory;
     }
 
-    static final class HybridDirectory extends NIOFSDirectory {
+    static final class HybridDirectory extends NIOFSDirectory implements DirectIOIndexInputSupplier {
         private final MMapDirectory delegate;
+        private final DirectIODirectory directIODelegate;
 
         HybridDirectory(LockFactory lockFactory, MMapDirectory delegate) throws IOException {
             super(delegate.getDirectory(), lockFactory);
             this.delegate = delegate;
+
+            DirectIODirectory directIO;
+            try {
+                directIO = new DirectIODirectory(delegate, 8192, DirectIODirectory.DEFAULT_MIN_BYTES_DIRECT) {
```

Comment:
Why do we need 8k and not a single page? Should we get this value from the system's page size?

Reply:
A 1024 dims x 4-bytes-per-float = 4k.  I suggested 8k so allow for unalignment across page boundaries.

Reply:
Yep that's fine but I wonder if a small buffer makes a difference here. If most reads are between 1k and 8k for instance, using the buffer twice to read 8k instead of once shouldn't make a difference. We still need to read two pages. 
Regarding alignement, we can also adapt the codec in a follow up to ensure it is aligned at the page level? That should be feasible for all dimensions that are a power of 2? For 1024 dimensions, it can be perfectly aligned to fit into a single page and we should be able to optimise for this case to never read more than one page? I mean that's the goal ultimately.

Reply:
That's something we could look at in a separate PR?

Reply:
Sure, we can find the optimal buffer size later, 8kb is a good compromise for now. Mind adding a comment to explain the rationale?